devicePointerLength gets set to -1 at any stream free in CUDA

```
@Override
    public boolean freeDevicePointer() {
        CUdeviceptr devicePointer = pointersToContexts.get(Thread.currentThread().getName());
    	if(devicePointer != null && !freed.get()) {
			allocated.addAndGet(-devicePointerLength);
	        log.trace("freeing {} bytes, total: {}", devicePointerLength, allocated.get());
	        checkResult(JCuda.cudaFree(devicePointer));
			devicePointer = null;
            freed.set(true);
            devicePointerLength = -1;
            pointersToContexts.remove(Thread.currentThread().getName());
			return true;	
    	} 
    	return false;
    }
```

The device pointer length should be associated with the device pointer in the stream, the length is being set to -1 and subsequent copy to hosts are throwing:

```
Caused by: jcuda.CudaException: INVALID CUresult: 11
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.checkResult(BaseCudaDataBuffer.java:340) ~[classes/:na]
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.copyToHost(BaseCudaDataBuffer.java:364) ~[classes/:na]
	at org.nd4j.linalg.jcublas.CublasPointer.copyToHost(CublasPointer.java:40) ~[classes/:na]
```

Will fix this ASAP.

--- a/nd4j-jcublas-parent/nd4j-jcublas-7.0/src/test/java/jcublas/kernel/TestMultipleThreads.java
+++ b/nd4j-jcublas-parent/nd4j-jcublas-7.0/src/test/java/jcublas/kernel/TestMultipleThreads.java
@@ -1,0 +1,55 @@
+package jcublas.kernel;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Test;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+
+public class TestMultipleThreads {
+	
+	@Test
+	public void testMultipleThreads() throws InterruptedException {
+		int numThreads = 50;
+		final INDArray array = Nd4j.rand(3000, 3000);
+		final INDArray expected = array.dup().mmul(array).mmul(array).div(array).div(array);
+		final AtomicInteger correct = new AtomicInteger();
+		final CountDownLatch latch = new CountDownLatch(numThreads);
+		
+		ExecutorService executors = Executors.newCachedThreadPool();
+		
+		for(int x = 0; x< numThreads; x++) {
+			executors.execute(new Runnable() {
+				@Override
+				public void run() {
+					try
+					{
+						int total = 10000;
+						int right = 0;
+						for(int x = 0; x<total; x++) {
+							INDArray actual = array.dup().mmul(array).mmul(array).div(array).div(array);
+							if(expected.equals(actual)) right++;						
+						}
+						
+						if(total == right)
+							correct.incrementAndGet();
+					} finally {
+						latch.countDown();
+					}
+					
+				}
+			});
+		}
+		
+		latch.await();
+		
+		assertEquals(numThreads, correct.get());
+		
+	}
+
+}

--- a/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/jcuda/utils/KernelLauncher.java
+++ b/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/jcuda/utils/KernelLauncher.java
@@ -1037,7 +1037,7 @@ public class KernelLauncher {
             }
         }
 
-
+        setContext();
 
         checkResult(cuLaunchKernel(function,
                 gridSize.x, gridSize.y, gridSize.z,

--- a/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java
+++ b/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/org/nd4j/linalg/jcublas/buffer/BaseCudaDataBuffer.java
@@ -58,10 +58,9 @@ import org.slf4j.LoggerFactory;
  */
 public abstract class BaseCudaDataBuffer implements JCudaBuffer {
 
-	protected transient long devicePointerLength;
 	protected transient Pointer hostPointer;
 	protected transient ByteBuffer hostBuffer;
-	protected transient Map<String,CUdeviceptr> pointersToContexts = new ConcurrentHashMap<>();
+	protected transient Map<String,DevicePointerInfo> pointersToContexts = new ConcurrentHashMap<>();
     protected AtomicBoolean modified = new AtomicBoolean(false);
     protected int length;
     protected int elementSize;
@@ -73,6 +72,29 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
     static AtomicLong totalAllocated = new AtomicLong();
     
     private static Logger log = LoggerFactory.getLogger(BaseCudaDataBuffer.class);
+    
+    /**
+     * Provides information about a device pointer 
+     * @author bam4d
+     *
+     */
+    private static class DevicePointerInfo {
+    	final private CUdeviceptr pointer;
+    	final private long length;
+    	
+    	public CUdeviceptr getPointer() { 
+    		return pointer;
+    	}
+    	
+    	public long getLength() {
+    		return length;
+    	}
+    	
+    	public DevicePointerInfo(CUdeviceptr pointer, long length) {
+    		this.pointer = pointer;
+    		this.length = length;
+    	}
+    }
     
     @Override
     public void setHostBuffer(ByteBuffer hostBuffer) {
@@ -176,20 +198,22 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
 
     @Override
 	public CUdeviceptr getDevicePointer() {
-        CUdeviceptr devicePointer = pointersToContexts.get(Thread.currentThread().getName());
+        DevicePointerInfo devicePointerInfo = pointersToContexts.get(Thread.currentThread().getName());
 
-    	if(devicePointer == null) {
-    		devicePointer = new CUdeviceptr();
-    		devicePointerLength = getElementSize() * length();
+    	if(devicePointerInfo == null) {
+    		CUdeviceptr devicePointer = new CUdeviceptr();
+    		int devicePointerLength = getElementSize() * length();
     		allocated.addAndGet(devicePointerLength);
     		totalAllocated.addAndGet(devicePointerLength);
     		log.trace("Allocating {} bytes, total: {}, overall: {}", devicePointerLength, allocated.get(), totalAllocated);
+    		
+    		devicePointerInfo = new DevicePointerInfo(devicePointer, devicePointerLength);
     		checkResult(JCuda.cudaMalloc(devicePointer, devicePointerLength));
-            pointersToContexts.put(Thread.currentThread().getName(),devicePointer);
+            pointersToContexts.put(Thread.currentThread().getName(),devicePointerInfo);
     		freed.set(false);
     	}
     	
-    	return devicePointer;
+    	return devicePointerInfo.getPointer();
     }
 
 
@@ -343,14 +367,13 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
     
     @Override
     public boolean freeDevicePointer() {
-        CUdeviceptr devicePointer = pointersToContexts.get(Thread.currentThread().getName());
-    	if(devicePointer != null && !freed.get()) {
-			allocated.addAndGet(-devicePointerLength);
-	        log.trace("freeing {} bytes, total: {}", devicePointerLength, allocated.get());
-	        checkResult(JCuda.cudaFree(devicePointer));
-			devicePointer = null;
+    	DevicePointerInfo devicePointerInfo = pointersToContexts.get(Thread.currentThread().getName());
+    	if(devicePointerInfo != null && !freed.get()) {
+			allocated.addAndGet(-devicePointerInfo.getLength());
+	        log.trace("freeing {} bytes, total: {}", devicePointerInfo.getLength(), allocated.get());
+	        checkResult(JCuda.cudaFree(devicePointerInfo.getPointer()));
+	        devicePointerInfo = null;
             freed.set(true);
-            devicePointerLength = -1;
             pointersToContexts.remove(Thread.currentThread().getName());
 			return true;	
     	} 
@@ -359,9 +382,9 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
     
     @Override
     public void copyToHost() {
-        CUdeviceptr devicePointer = pointersToContexts.get(Thread.currentThread().getName());
-        if(devicePointer != null)
-    		checkResult(JCuda.cudaMemcpy(hostPointer, devicePointer, devicePointerLength, cudaMemcpyKind.cudaMemcpyDeviceToHost));
+    	DevicePointerInfo devicePointerInfo = pointersToContexts.get(Thread.currentThread().getName());
+        if(devicePointerInfo != null)
+    		checkResult(JCuda.cudaMemcpy(hostPointer, devicePointerInfo.getPointer(), devicePointerInfo.getLength(), cudaMemcpyKind.cudaMemcpyDeviceToHost));
     }
 
 
@@ -453,7 +476,6 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
 
         BaseCudaDataBuffer that = (BaseCudaDataBuffer) o;
 
-        if (devicePointerLength != that.devicePointerLength) return false;
         if (length != that.length) return false;
         if (elementSize != that.elementSize) return false;
         if (isPersist != that.isPersist) return false;
@@ -470,8 +492,7 @@ public abstract class BaseCudaDataBuffer implements JCudaBuffer {
 
     @Override
     public int hashCode() {
-        int result = (int) (devicePointerLength ^ (devicePointerLength >>> 32));
-        result = 31 * result + (hostPointer != null ? hostPointer.hashCode() : 0);
+        int result = (int) (hostPointer != null ? hostPointer.hashCode() : 0);
         result = 31 * result + (hostBuffer != null ? hostBuffer.hashCode() : 0);
         result = 31 * result + (pointersToContexts != null ? pointersToContexts.hashCode() : 0);
         result = 31 * result + (modified != null ? modified.hashCode() : 0);

--- a/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/org/nd4j/linalg/jcublas/context/ContextHolder.java
+++ b/nd4j-jcublas-parent/nd4j-jcublas-common/src/main/java/org/nd4j/linalg/jcublas/context/ContextHolder.java
@@ -78,15 +78,19 @@ public class ContextHolder {
     	
     	if(stream == null) {
     		stream = new CUstream();
-    		int result = JCudaDriver.cuStreamCreate(stream, CUstream_flags.CU_STREAM_DEFAULT);
-    		if (result != CUresult.CUDA_SUCCESS) {
-                throw new CudaException("Failed to create a stream: "+ CUresult.stringFor(result));
-            }
+    		checkResult(JCudaDriver.cuCtxSetCurrent(ctx));
+    		checkResult(JCudaDriver.cuStreamCreate(stream, CUstream_flags.CU_STREAM_DEFAULT));
     		contextStreams.put(ctx, currentThread.getName(), stream);
     	}
     	
     	return stream;
     }
+
+	private void checkResult(int result) {
+		if (result != CUresult.CUDA_SUCCESS) {
+		    throw new CudaException("Failed to create a stream: "+ CUresult.stringFor(result));
+		}
+	}
 
     /**
      * Retrieve a context for use with the current thread

